mongoengine/base/datastructures.py

```diff
@@ -88,11 +88,16 @@ def update(self, *args, **kwargs):
         return super(BaseDict, self).update(*args, **kwargs)
 
     def _mark_as_changed(self, key=None):
-        if hasattr(self._instance, '_mark_as_changed'):
-            if key:
-                self._instance._mark_as_changed('%s.%s' % (self._name, key))
-            else:
-                self._instance._mark_as_changed(self._name)
+        try:
+            if hasattr(self._instance, '_mark_as_changed'):
+                if key:
+                    self._instance._mark_as_changed(
+                        '%s.%s' % (self._name, key % len(self))
+                    )
+                else:
+                    self._instance._mark_as_changed(self._name)
+        except ReferenceError:
+            pass
```

Comment:
this was a hack to fix this failing test

`======================================================================
ERROR: test_regular_groups_type (bcore.apps.teamdiscussion.tests.TestTeamDiscussionGroupTypes)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/ubuntu/hbx/bcore/bcore/apps/teamdiscussion/tests.py", line 159, in test_regular_groups_type
    follow=True
  File "/home/ubuntu/hbx/python/lib/python3.5/site-packages/django/test/client.py", line 515, in post
    secure=secure, **extra)
  File "/home/ubuntu/hbx/python/lib/python3.5/site-packages/django/test/client.py", line 314, in post
    secure=secure, **extra)
  File "/home/ubuntu/hbx/python/lib/python3.5/site-packages/django/test/client.py", line 380, in generic
    return self.request(**r)
  File "/home/ubuntu/hbx/python/lib/python3.5/site-packages/django/test/client.py", line 467, in request
    six.reraise(*exc_info)
  File "/home/ubuntu/hbx/python/lib/python3.5/site-packages/django/utils/six.py", line 686, in reraise
    raise value
  File "/home/ubuntu/hbx/python/lib/python3.5/site-packages/django/core/handlers/base.py", line 149, in get_response
    response = self.process_exception_by_middleware(e, request)
  File "/home/ubuntu/hbx/python/lib/python3.5/site-packages/django/core/handlers/base.py", line 147, in get_response
    response = wrapped_callback(request, *callback_args, **callback_kwargs)
  File "/home/ubuntu/hbx/python/lib/python3.5/site-packages/django/views/decorators/csrf.py", line 58, in wrapped_view
    return view_func(*args, **kwargs)
  File "/home/ubuntu/hbx/python/lib/python3.5/site-packages/tastypie/resources.py", line 219, in wrapper
    response = callback(request, *args, **kwargs)
  File "/home/ubuntu/hbx/python/lib/python3.5/site-packages/tastypie/resources.py", line 450, in dispatch_list
    return self.dispatch('list', request, **kwargs)
  File "/home/ubuntu/hbx/python/src/django-tastypie-mongoengine/tastypie_mongoengine/resources.py", line 431, in dispatch
    return self._wrap_request(request, lambda: super(MongoEngineResource, self).dispatch(request_type, request, **kwargs))
  File "/home/ubuntu/hbx/python/src/django-tastypie-mongoengine/tastypie_mongoengine/resources.py", line 394, in _wrap_request
    return fun()
  File "/home/ubuntu/hbx/python/src/django-tastypie-mongoengine/tastypie_mongoengine/resources.py", line 431, in <lambda>
    return self._wrap_request(request, lambda: super(MongoEngineResource, self).dispatch(request_type, request, **kwargs))
  File "/home/ubuntu/hbx/python/lib/python3.5/site-packages/tastypie/resources.py", line 482, in dispatch
    response = method(request, **kwargs)
  File "/home/ubuntu/hbx/python/lib/python3.5/site-packages/tastypie/resources.py", line 1384, in post_list
    updated_bundle = self.obj_create(bundle, **self.remove_api_resource_names(kwargs))
  File "/home/ubuntu/hbx/bcore/bcore/apps/teamdiscussion/api.py", line 80, in obj_create
    bundle, activity_origin, user_id, course_instance_id, cohort_id, reply_type, tei_id, text
  File "/home/ubuntu/hbx/bcore/bcore/apps/teamdiscussion/models.py", line 143, in create_reply
    TeamDiscussionNotification(discussion_object, group_users).broadcast()
  File "/home/ubuntu/hbx/bcore/bcore/apps/notifications/notifications/team_discussion_notification.py", line 21, in __init__
    super(TeamDiscussionNotification, self).__init__()
  File "/home/ubuntu/hbx/bcore/bcore/apps/notifications/notifications/notification.py", line 20, in __init__
    self.users = self.set_subscribers()
  File "/home/ubuntu/hbx/bcore/bcore/apps/notifications/notifications/team_discussion_notification.py", line 24, in set_subscribers
    self.group_users.remove(self.notification_instance.user_id)
  File "/home/ubuntu/hbx/python/lib/python3.5/site-packages/mongoengine/base/datastructures.py", line 187, in remove
    self._mark_as_changed()
  File "/home/ubuntu/hbx/python/lib/python3.5/site-packages/mongoengine/base/datastructures.py", line 199, in _mark_as_changed
    if hasattr(self._instance, '_mark_as_changed'):
ReferenceError: weakly-referenced object no longer exists`